### PR TITLE
Create a build persistence layer

### PR DESCRIPTION
**Pull request by zwebb**
In PR #4, some code was added to locally persist builds so that time statistics could be calculated over a longer period than what Jenkins kept. This code has now been moved out to the `jenkins_persistence.py` file. The remaining code in `butlercam.py` has been refactored to always access builds through this layer, meaning that the entire app can now make use of the cached builds.

In addition, a number of new unit tests were written. Every function in `butlercam.py` that doesn't rely directly on Jenkins access is now being covered in `butlercam_tests.py`.

### CODE REVIEW

File: flask/butlercam.py [hunk 1]
```diff
@@ -77,20 +72,25 @@ def show_pipeline(pipeline):
         latest_build_status = get_last_complete_status(PIPELINES[pipeline][1])
         building = True
 
-    csv_data = generate_csvs(historical_builds)
+    latest_keys = all_builds.keys()[-20:]
+    latest_builds = OrderedDict()
+    for i in latest_keys:
+        latest_builds[i] = all_builds[i]
+
+    csv_data = generate_build_csv(all_builds)
```
Comment (technolo-g) [inline flask/butlercam.py]: This appears to be run on all requests, is that correct? If so, we should probably move this to a different route so it is hit less frequently. Perhaps we can add a route like `/pipeline/<pipeline>/csv` to generate and display this view?

File: flask/jenkins_persistence.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+"""
+A component that queries Jenkins for builds and saves them in a local
+Redis database so that they can be accessed even after Jenkins has
+stopped holding onto them.
+"""
+# pylint: disable=C0103
```
Comment (technolo-g) [inline flask/jenkins_persistence.py]: Why do we have to disable this?

File: flask/jenkins_persistence.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+app = Flask(__name__)
+app.config.from_object(default_settings)
+if 'BUTLERCAM_SETTINGS_FILE' in os.environ:
+    app.config.from_envvar('BUTLERCAM_SETTINGS_FILE')
+PIPELINES = app.config['PIPELINES']
+
+db = Database(host=app.config['REDIS_HOST'], db=0)
+cache = db.cache(default_timeout=app.config['CACHE_TIMEOUT'])
+
+continuous = True
```
Comment (technolo-g) [inline flask/jenkins_persistence.py]: Can we document this?

File: flask/jenkins_persistence.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+def save_builds(pipeline, builds):
+    """
+    Given the name of a pipeline and a dict of builds, clear the
+    existing saved builds for that pipeline and add all of the given
+    builds in their place.
+    """
+
+    redis_builds = db.Array(pipeline + ':builds')
+    clear_saved_builds(pipeline)
```
Comment (technolo-g) [inline flask/jenkins_persistence.py]: Why do we have to do this?

File: flask/jenkins_persistence.py (hunk 2 above)
Comment (zwebb) [inline flask/jenkins_persistence.py]: Travis was giving me errors that I didn't really understand:

```
C:  8, 0: Invalid constant name "app" (invalid-name)
C: 14, 0: Invalid constant name "db" (invalid-name)
C: 15, 0: Invalid constant name "cache" (invalid-name)
C: 17, 0: Invalid constant name "continuous" (invalid-name)
```

I saw that `butlercam.py` had that same disable statement at the top, so I copied it over and it seemed to fix the errors.

File: flask/jenkins_persistence.py (hunk 4 above)
Comment (zwebb) [inline flask/jenkins_persistence.py]: I wanted it to be possible for the saved builds to be updated if their information changed--for example, if they were first caught in the "building" state, and later resolved to a "success" or something. I thought this was a simpler solution than checking each one to see if the Jenkins copy is different from the Redis copy.

File: flask/jenkins_persistence.py (hunk 3 above)
Comment (zwebb) [inline flask/jenkins_persistence.py]: Turns out this wasn't needed in this file at all. I went ahead and added a comment above it on `butlercam.py`.

**Pull request merged**